maven.indexer/src/org/netbeans/modules/maven/indexer/ClassDependencyIndexCreator.java

```diff
@@ -254,45 +331,52 @@ private static void addDependenciesToMap(String referrer, byte[] data, Map<Strin
                     continue;
                 }
                 String clazz = name.substring(0, name.length() - 6);
-                ByteArrayOutputStream baos = new ByteArrayOutputStream(Math.max((int) entry.getSize(), 0));
-                InputStream is = jf.getInputStream(entry);
-                try {
-                    FileUtil.copy(is, baos);
-                } finally {
-                    is.close();
+                classNames.add(clazz);
+            }
+            e = jf.entries();
+            while (e.hasMoreElements()) {
+                JarEntry entry = e.nextElement();
+                String name = entry.getName();
+                if (!name.endsWith(".class")) {
+                    continue;
+                }
+                int size = Math.max((int) entry.getSize(), 0);
+                if (size > 0) {
+                    // Parsing is considerably faster if the data is preloaded
+                    // into a byte array, likely due to random access
+                    byte[] target = bytes(size);
+                    try (InputStream in = jf.getInputStream(entry)) {
+                        for (int pos = 0, count = in.read(target);
+                                count != -1 && pos < size;
+                                count = in.read(target, pos, size - pos), pos += count == -1 ? 0 : count);
```

Comment:
I'm fine with this.  In tests, I never encountered the condition with JarEntry, but if you have, that's a reason to change it.  Want to send a pull request to my fork?

Reply:
Re using ClassLoader.getResource():  I don't remember the final outcome of discussions of using that on class files in JDK 9's module system (I had to leave the expert group when I rejoined Oracle for a year in 2015), but where it was leaning then was that it was not going to work.  So I wouldn't see using that mechanism as advisable.

Copy/pasting my response to Emilian from above, re using that as a test:

It is not so much that the class list is infallible, as that, in practice, even if the test misses a JDK class, it will not result in incorrect behavior.

The test exists, basically, because you do not want to fill the index with all uses of, say java.lang.Integer - in practice nobody searches for that, and if they did there would be too many results for it to be useful.

So, what are the possible failure scenarios:

A new package was added to the JDK, and gets indexed: That's fine and will be for many years - even if it eventually is used as heavily as java.lang.Integer, it isn't right now, and there are years to add the package to the test before it's used enough to cause serious index bloat.

We are indexing a library downloaded from Maven which adds classes to a JDK package. First, in practice, that doesn't happen - the main reasons it might are malicious, and in fact, one of the things JDK 9's module system is there to prevent. Second, it means paying a performance penalty to handle a corner-case that, that 99.99999% of Maven users will never encounter (and if they do, it's because they downloaded a library that contains malicious code).

Neither of those are cases that justify a more expensive test.

The package membership test is sufficient.

Re performance comments:  [Here is a test](https://timboudreau.com/files/PerfTest.java) you can download and add to the unit test, which will scan `~/.m2/repository`.  To get meaningful results, the code needs to index enough JARs to take JIT compiling out of the picture, and you need to run it twice in a row in each case and look at values from the second run, so that you're not measuring the overhead of your OS populating your disk cache.

As I noted, the major benefit is likely to be reduced GC pressure, and where that benefit is going to show up is in a loaded JVM which is doing indexing in the background, while the IDE needs to remain responsive to the user.  A test that measures indexing performance in isolation does not measure that.